[MC/DC][Coverage] Enable profile correlation for MC/DC

belyaevrd opened the pull request:
When using the `-fcoverage-mcdc` option in profile correlation mode MC/DC coverage is not actually collected.

In binary profile correlation mode this completes elements of a vector of per-function profile data structures called `Data` with BitmapPtr and NumBitmapBytes values that are taken from profile data section in the same way as it is done for Counters.

In debug info correlation mode this adds new `Profile Bitmap Type` DIEs to DWARFContext. These entries contain FunctionName and NumBitmapBits for functions. They are used by `correlateProfileDataImpl()` function to obtain BitmapPtr and NumBitmapBytes values to complete the corresponding elements of the vector of per-function profile data structures called `Data`. Creating and reading these new DIEs occur in the same way as it is done for DIEs of the type `Profile Data Type`.

Fixes #97385

evodius96 commented:
Hello -- thank you for implementing this!  I can review, if you like, but it would probably also be good to include @ellishg as well.

evodius96 reviewed (approved):
Most of the changes appear to be straightforward, but I would like @ellishg to look at `InstrProfCorrelator.cpp`.  Other than that, LGTM.

ellishg commented on llvm/lib/ProfileData/InstrProfCorrelator.cpp:
```diff
@@ -342,106 +358,199 @@ bool DwarfInstrProfCorrelator<IntPtrT>::isDIEOfProbe(const DWARFDie &Die) {
-      if (auto Err = AnnotationNameOrErr.takeError()) {
-        consumeError(std::move(Err));
-        continue;
+    if (Prefix == getInstrProfCountersVarPrefix()) {
```
This code is getting large. Can we factor these two cases into separate functions?

ellishg reviewed (commented):
Looks good overall. Hopefully we can refactor some changes to `InstrProfCorrelator.cpp` because the functions are getting quite large.

Sorry for dropping the ball on reviewing this. Next time feel free to ping me again :)

belyaevrd committed:
[MC/DC][Coverage] Enable profile correlation for MC/DC

When using the `-fcoverage-mcdc` option in profile correlation mode MC/DC
coverage is not actually collected.

In binary profile correlation mode this completes elements of a vector of
per-function profile data structures called `Data` with BitmapPtr and
NumBitmapBytes values that are taken from profile data section in the same way
as it is done for Counters.

In debug info correlation mode this adds new `Profile Bitmap Type` DIEs to
DWARFContext. These entries contain FunctionName and NumBitmapBits for
functions. They are used by `correlateProfileDataImpl()` function to obtain
BitmapPtr and NumBitmapBytes values to complete the corresponding elements of
the vector of per-function profile data structures called `Data`. Creating and
reading these new DIEs occur in the same way as it is done for DIEs of the type
`Profile Data Type`.

Fixes #97385

belyaevrd commented:
@ellishg Sorry for the long delay. I’ve updated the patch based on your comments. Could you take a look, please?

ellishg commented on compiler-rt/test/profile/instrprof-mcdc-correlation.c:
```diff
@@ -0,0 +1,21 @@
+// REQUIRES: linux || darwin
+// Default
+// RUN: %clang -o %t.normal -fprofile-instr-generate -fcoverage-mapping -fcoverage-mcdc %S/Inputs/instrprof-mcdc-correlation.cpp
```
Why does `instrprof-mcdc-correlation.cpp` need to be separate from `instrprof-mcdc-correlation.c‎`? Let's either move the run lines into the `.cpp` file or the c code in here.

belyaevrd replied:
Done

ellishg commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -1655,6 +1655,39 @@ InstrLowerer::getOrCreateRegionBitmaps(InstrProfMCDCBitmapInstBase *Inc) {
+          MDNode::get(Ctx, FunctionNameAnnotation),
+          MDNode::get(Ctx, NumBitmapBitsAnnotation),
+      });
+      auto *DICounter = DB.createGlobalVariableExpression(
```
Can you add an LLVM lit test for this debug info? We have a similar test here

https://github.com/llvm/llvm-project/blob/1bbfade93ff3ea1df295a740ef5547763304f891/llvm/test/Instrumentation/InstrProfiling/debug-info-correlate.ll#L17-L19

belyaevrd replied:
Done

ellishg commented on llvm/lib/ProfileData/InstrProfCorrelator.cpp:
```diff
@@ -297,29 +315,26 @@ Error InstrProfCorrelatorImpl<IntPtrT>::dumpYaml(int MaxWarnings,
+    uint32_t NumBitmapBytes) {
   // Check if a probe was already added for this counter offset.
-  if (!CounterOffsets.insert(CounterOffset).second)
+  if (CounterOffset && !CounterOffsets.insert(CounterOffset).second)
```
Why is this extra check needed? It's been a while since I've worked in this file. Is `CounterOffset == 0` a valid value?

belyaevrd replied:
Yes, `CounterOffset == 0` is a valid value. I tried not to skip entries that contain only Bitmap information, but actually this check was wrong. I have fixed this now.

ellishg commented on llvm/lib/ProfileData/InstrProfCorrelator.cpp:
```diff
@@ -361,102 +377,223 @@ bool DwarfInstrProfCorrelator<IntPtrT>::isDIEOfProbe(const DWARFDie &Die) {
+        return;
       }
-      return;
     }
```
This is new. Can you elaborate on why we are searching for a probe with the same name? This might need to be a separate PR since it impact all debug info correlation, not just MC/DC. Also, if we really need to search existing probes, we should consider using a map from function name to probe.

belyaevrd replied:
We need to merge data from function Counter and Bitmap DIEs and put it into one Probe. I think that it is actually optimal to do this in a single pass through the loop of all DIEs. I rewrote this using a map from FunctionName to (Probe, FunctionPtr) pair. Now it should work faster.

ellishg commented on llvm/lib/ProfileData/InstrProfCorrelator.cpp:
```diff
@@ -361,102 +377,223 @@ bool DwarfInstrProfCorrelator<IntPtrT>::isDIEOfProbe(const DWARFDie &Die) {
+        return;
       }
-      return;
     }
```
Same here. I believe this for loop makes this O(n^2)

ellishg commented:
Please try not to force push, which makes reviewing the latest changes difficult. Instead, create new commits and push them to the branch. When the PR is merged, github will automatically squash the commits into one to keep the history clean.

https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes

belyaevrd commented:
> Please try not to force push, which makes reviewing the latest changes difficult. Instead, create new commits and push them to the branch. When the PR is merged, github will automatically squash the commits into one to keep the history clean.
> 
> https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes

My bad, sorry for this.

ellishg commented on llvm/lib/ProfileData/InstrProfCorrelator.cpp:
```diff
@@ -361,102 +380,211 @@ bool DwarfInstrProfCorrelator<IntPtrT>::isDIEOfProbe(const DWARFDie &Die) {
+  auto MaybeAddProbe = [&](DWARFDie Die) {
+    if (isDIEOfProbe(Die, getInstrProfCountersVarPrefix()))
+      addCountersToDataProbe(Data, Probes, Die, UnlimitedWarnings,
+                             NumSuppressedWarnings);
```
I think we can get make `addCountersToDataProbe()` and `addBitmapToDataProbe()` return a `InstrProfCorrelator::Probe` instead of directly adding it to `Probes`. That way we can reduce code duplication to add to `Probes` in both functions. Also, I think we can avoid passing `Data` to these functions and do `this->NamesVec.push_back(*FunctionName);` on line 599 instead.

ellishg replied:
To be concrete, I'm imagining something like this

```
if (isDIEOfProbe(Die, getInstrProfCountersVarPrefix()))
  Probe = addCountersToDataProbe()
else if (isDIEOfProbe(Die, getInstrProfBitmapVarPrefix()))
  Probe = addBitmapToDataProbe()

// TODO: Need to figure out how to grab FunctionPtr
auto [It, WasInserted] = Probes.try_emplace(Probe.FunctionName, Probe, FunctionPtr);
if (!WasInserted) {
  auto *Probe = It->Second.first;
  // Fill out probe based on if this is a counter or bitmap die
  Probe.LinkageName = ...
}
```

ellishg commented on llvm/lib/ProfileData/InstrProfCorrelator.cpp:
```diff
@@ -361,102 +380,211 @@ bool DwarfInstrProfCorrelator<IntPtrT>::isDIEOfProbe(const DWARFDie &Die) {
+    Iter->second.second = *FunctionPtr;
+    return;
+  }
+  Probes[std::string(*FunctionName)] = {P, *FunctionPtr};
```
The neat thing about `StringMap` is that we can pass `StringRef`s without worrying about lifetimes. I think it copies the string's contents regardless.
```suggestion
  Probes[StringRef(*FunctionName)] = {P, *FunctionPtr};
```

belyaevrd replied:
If I understand correctly, `MapVector` keeps an internal vector of <KeyT, ValueT> pairs. If KeyT is `StringRef`, it can contain refs to dead strings. The third parameter of `MapVector` (equals to `StringMap`) changes only the internal map type organization. So, we need to store strings in the vector to ensure their lifetime during iteration.

ellishg reviewed (commented):
> > Please try not to force push, which makes reviewing the latest changes difficult. Instead, create new commits and push them to the branch. When the PR is merged, github will automatically squash the commits into one to keep the history clean.
> > https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes
> 
> My bad, sorry for this.

No worries!

Looks better, just have a few more comments. Also, it looks like `instrprof-debug-info-correlate-warnings.ll` is failing?

belyaevrd committed:
Insert new entries into `Probes` by returning the data from `addCountersToDataProbe`() and `addBitmapToDataProbe`().

ellishg commented on llvm/lib/ProfileData/InstrProfCorrelator.cpp:
```diff
@@ -465,6 +587,20 @@ void DwarfInstrProfCorrelator<IntPtrT>::correlateProfileDataImpl(
     for (const auto &Entry : CU->dies())
       MaybeAddProbe(DWARFDie(CU.get(), &Entry));
 
+  for (auto Iter = Probes.begin(); Iter != Probes.end(); ++Iter) {
```
Does this compile?
```suggestion
  for (const auto &[FunctionName, P, FunctionPtr] : Probes) {
```

belyaevrd replied:
I couldn't compile this, but the following code works. Should I fix the loop this way?
```suggestion
  for (const auto &[FunctionName, ProbeData] : Probes) {
    const auto &[Probe, FunctionPtr] = ProbeData;
```

ellishg replied:
Yeah that's fine!

belyaevrd replied:
Done

ellishg commented on llvm/lib/ProfileData/InstrProfCorrelator.cpp:
```diff
@@ -465,6 +587,20 @@ void DwarfInstrProfCorrelator<IntPtrT>::correlateProfileDataImpl(
+                         Probe.NumBitmapBytes);
+    }
+  }
+  Probes.clear();
```
I believe this isn't necessary because the function exits anyway.
```suggestion
```

ellishg reviewed (approved):
LGTM, just some minor nits. Thanks for working on this!

belyaevrd committed:
Remove the unnecessary clear() call at the end of the function

belyaevrd committed:
Switch to range-based for loop

belyaevrd commented:
Thanks for the reviews. I will merge today if there are no more comments.

belyaevrd committed:
Merge branch 'main' into mcdc-profile-correlation

belyaevrd commented:
Ah, I don't have write access. Could someone merge it for me, please?

ellishg commented:
> Ah, I don't have write access. Could someone merge it for me, please?

I can merge Monday morning unless someone else wants to first.

jroelofs commented:
Looks like the new test broke in GreenDragon, mind taking a look?

https://green.lab.llvm.org/job/clang-stage1-RA-as/job/main/1896

jroelofs commented:
```
# RUN: at line 11
llvm-profdata merge -o /Users/ec2-user/jenkins/workspace/llvm.org_clang-stage1-RA-as_main/clang-build/runtimes/runtimes-bins/compiler-rt/test/profile/Profile-arm64/Output/instrprof-mcdc-correlation.c.tmp-1.profdata --binary-file=/Users/ec2-user/jenkins/workspace/llvm.org_clang-stage1-RA-as_main/clang-build/runtimes/runtimes-bins/compiler-rt/test/profile/Profile-arm64/Output/instrprof-mcdc-correlation.c.tmp-1.exe /Users/ec2-user/jenkins/workspace/llvm.org_clang-stage1-RA-as_main/clang-build/runtimes/runtimes-bins/compiler-rt/test/profile/Profile-arm64/Output/instrprof-mcdc-correlation.c.tmp-1.profraw
# executed command: llvm-profdata merge -o /Users/ec2-user/jenkins/workspace/llvm.org_clang-stage1-RA-as_main/clang-build/runtimes/runtimes-bins/compiler-rt/test/profile/Profile-arm64/Output/instrprof-mcdc-correlation.c.tmp-1.profdata --binary-file=/Users/ec2-user/jenkins/workspace/llvm.org_clang-stage1-RA-as_main/clang-build/runtimes/runtimes-bins/compiler-rt/test/profile/Profile-arm64/Output/instrprof-mcdc-correlation.c.tmp-1.exe /Users/ec2-user/jenkins/workspace/llvm.org_clang-stage1-RA-as_main/clang-build/runtimes/runtimes-bins/compiler-rt/test/profile/Profile-arm64/Output/instrprof-mcdc-correlation.c.tmp-1.profraw
# .---command stderr------------
# | warning: BitmapPtr out of range for function: Actual=0xc000 Expected=[0xc000, 0xc003) at data offset=0x0
# | warning: /Users/ec2-user/jenkins/workspace/llvm.org_clang-stage1-RA-as_main/clang-build/runtimes/runtimes-bins/compiler-rt/test/profile/Profile-arm64/Output/instrprof-mcdc-correlation.c.tmp-1.profraw: malformed instrumentation profile data: bitmap offset 31525193096626176 is greater than the maximum bitmap offset 2
# | error: no profile can be merged
# `-----------------------------
# error: command failed with exit status: 1
```

ellishg commented:
@belyaevrd could you take a look at this error?

belyaevrd commented:
> @belyaevrd could you take a look at this error?

Yes, I’m on it.

zmodem commented:
I've reverted in the meantime.

We hit this also, here are our logs:
- https://logs.chromium.org/logs/chromium/buildbucket/cr-buildbucket/8681377437372285233/+/u/package_clang/stdout
- https://logs.chromium.org/logs/chromium/buildbucket/cr-buildbucket/8681377437372285217/+/u/package_clang/stdout
